client/pool.c

```diff
@@ -1,417 +1,680 @@
+				}
+				memcpy(conn_data->miner->id.data, conn_data->data, sizeof(struct xdag_field));	//TODO:do I need to copy whole field?
+			}
+
+			xdag_hash_t hash;
+			xdag_hash_final(task->ctx0, conn_data->data, sizeof(struct xdag_field), hash);
+			xdag_set_min_share(task, conn_data->miner->id.data, hash);
+			calculate_nopaid_shares(conn_data, task, hash);
 		}
+	}
+
+	return 1;
+}
 
-		if (!topay) continue;
-		
-		memcpy(fields[nfld].data, m->id.data, sizeof(xdag_hashlow_t));
-		fields[nfld].amount = topay;
-		fields[0].amount += topay;
-		
-		xdag_log_xfer(fields[0].data, fields[nfld].data, topay);
-		
-		if (++nfld == nfields) {
-			xdag_create_block(fields, 1, nfld - 1, 0, 0, NULL);
-			nfld = 1;
-			fields[0].amount = 0;
+static int send_data_to_connection(connection_list_element *connection, int index)
+{
+	struct xdag_field data[2];
+	int nfld = 0;
+	struct connection_pool_data *conn_data = &connection->connection_data;
+
+	uint64_t task_index = g_xdag_pool_task_index;
+	struct xdag_pool_task *task = &g_xdag_pool_task[task_index & 1];
+
+	if(conn_data->task_index < task_index) {
+		conn_data->task_index = task_index;
+		//m->shares_count = 0;
+		nfld = 2;
+		memcpy(data, task->task, nfld * sizeof(struct xdag_field));
+	} else if(!conn_data->balance_sent && conn_data->miner && time(0) >= (conn_data->task_time << 6) + 4) {
+		conn_data->balance_sent = 1;
+		memcpy(data[0].data, conn_data->miner->id.data, sizeof(xdag_hash_t));
+		data[0].amount = xdag_get_balance(data[0].data);
+		nfld = 1;
+	}
+
+	if(nfld) {
+		for(int j = 0; j < nfld; ++j) {
+			dfslib_encrypt_array(g_crypt, (uint32_t*)(data + j), DATA_SIZE, conn_data->nfield_out++);
 		}
+
+		int length = write(conn_data->connection_descriptor.fd, (void*)data, nfld * sizeof(struct xdag_field));
+
+		if(length != nfld * sizeof(struct xdag_field)) {
+			close_connection(connection, index, "write error");
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+void *pool_main_thread(void *arg)
+{
+	connection_list_element *elt, *eltmp;
+
+	while(!g_xdag_sync_on) {
+		sleep(1);
 	}
 
-    if(nfld > 1) {
-        xdag_create_block(fields, 1, nfld - 1, 0, 0, NULL);
-    }
+	for(;;) {
+		pthread_mutex_lock(&g_descriptors_mutex);
+		const int miners_count = g_connections_count;
+		int res = poll(g_fds, miners_count, 1000);
+		pthread_mutex_unlock(&g_descriptors_mutex);
+
+		if(!res) continue;
+
+		int index = 0;
+		LL_FOREACH_SAFE(g_connection_list_head, elt, eltmp)
+		{
+			struct pollfd *p = g_fds + index++;
+
+			if(p->revents & POLLNVAL) continue;
+
+			if(p->revents & POLLHUP) {
+				close_connection(elt, index, "socket hangup");
+				continue;
+			}
+
+			if(p->revents & POLLERR) {
+				close_connection(elt, index, "socket error");
+				continue;
+			}
+
+			if(p->revents & POLLIN) {
+				if(!recieve_data_from_connection(elt, index)) {
+					continue;
+				}
+			}
+
+			if(p->revents & POLLOUT) {
+				if(!send_data_to_connection(elt, index)) {
+					continue;
+				}
+			}
+		}
+
+		sleep(1);
```

Comment:
If sleep here, it means the pool only can handle 1000 connections per second. It would be a performance issue.

Reply:
poll(g_fds, miners_count, 1000) - count of connections is miners_count. 1000 - tomeout in ms.

Reply:
About sleep: we did some I/O with network, now we can wait a bit.
But original code had condition
if(!done) sleep(1);
it means we sleep only when nothing was processed during last iteration.
Maybe it makes sense to return this condotion. What do you think?

Reply:
Yes, if(!done) sleep(1) is better, when nothing done sleep for a while. If have I/O on different fd, it will ok without sleep, since read/write is synchronized method.

Reply:
Ok

Reply:
Done.